### PR TITLE
PARQUET-2131:  Number values decoded DCHECKs should be exceptions

### PR DESCRIPTION
**Pull request by tachyonwill**
As discussed on some other bugs, there are some parquet-cpp DCHECKs on
the number of values decoded that really should be exceptions. When
invalid Parquet files are read, it is possible for the decoders to
return less values than expected and this should be signaled back to
the user even in non-debug mode and it should not be a crash in debug mode.

### CODE REVIEW

**Comment by tachyonwill** [comment]
Fuzzer found examples: https://github.com/apache/arrow-testing/pull/78

File: cpp/src/parquet/column_reader.cc [hunk 1]
```diff
@@ -1478,13 +1478,21 @@ class TypedRecordReader : public ColumnReaderImplBase<DType>,
     int64_t num_decoded = this->current_decoder_->DecodeSpaced(
         ValuesHead<T>(), static_cast<int>(values_with_nulls),
         static_cast<int>(null_count), valid_bits, valid_bits_offset);
-    DCHECK_EQ(num_decoded, values_with_nulls);
+    if (ARROW_PREDICT_FALSE(num_decoded != values_with_nulls)) {
+      ParquetException::EofException("Decoded values " + std::to_string(num_decoded) +
+                                     " does not match expected " +
+                                     std::to_string(values_with_nulls));
+    }
```
Comment (pitrou) [inline cpp/src/parquet/column_reader.cc:1485]: Can you factor out the check and error message in a convenience function or method?

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks for the PR. Can you update the testing submodule now that the corresponding PR has been merged?

File: cpp/src/parquet/column_reader.cc (hunk 1 above)
Comment (tachyonwill) [reply cpp/src/parquet/column_reader.cc]: Done

**Review by emkornfield — APPROVED** [review]

**Comment by tachyonwill** [comment]
Merging this should wait until https://github.com/apache/arrow/pull/12532 is merged since the submodule update pulled in the test files for that.

**Comment by emkornfield** [comment]
Will merge on green CI.

**Comment by emkornfield** [comment]
+1, failures look unrelated.

**Pull request closed without merging**